### DISCUSSION TITLE
Euctp (ICS2) support for phase4

### DISCUSSION THREAD

**Discussion by ulrikian** [phax/phase4#222] [General]
Hello,
we are currently starting to add support for the ICS2 customs declarations which use AS4 to phase4
https://taxation-customs.ec.europa.eu/customs-4/customs-security/import-control-system-2-ics2-0/faq_en
SD-ICS2-HTI-Interface Control Document-SfA-v3.30.pdf

we talked to our customer and he was fine with contributing to an open source project.

Our current implementation idea is:
- implement the pmode as module phase4-profile-euctp (base on the info in the pdf) 
- implement a client as module phase4-euctzp-client 
- implement a server based on springboot 

Is this ok for you?
Can you give us any leads on a good starting base? We are mostly done with the pmode and just about to start on the client.
Which would be a good base for the client? phase-4-bdew-client , phase4-cef-client, phase4-peppol-client?

Best regards, 
uli
us@codecare.de +49 176 8426 5352

**Comment by phax** [comment]
Hi Uli,
Thank you, that sounds very promising and I am happily looking forward to your contribution!
* Regarding the PMode, as this is an EU funded project, I assume the "CEF" based components are good starting points.
* For the client, the cef and the bdew clients are pretty thin layers and just add some common functionality, where the peppol-client has a lot more business logic in it (like support for validation, SMP lookup etc.). As I don't know the details of the ICS2 profile yet, I cannot make a good judgement call. As I am in doubt, I would recommend to start small and look at the cef client.
* For the server there are 2 different approaches: as phase4 is in general "thirdparty framework agnostic" and therefore runs without Spring and without the Servlet specification as well, I see the Spring Boot stuff really just as a demo. However, as I also saw people struggling with the multi module Maven project usage, I extracted a totally separate project https://github.com/phax/phase4-peppol-standalone that people can use as a kick-start. However, that project of course has less visibility compared to the main project.

So hope that answers your question.

Additionally, I sent you an invite to my Slack channel (phoss.slack.com) and encourage other interested parties to join as well - that also worked pretty okay for the BDEW implementation etc.

hth, Philip

**Reply by phax** [reply]
And btw. don't put too much effort in the coding style - I know it is different and I will take care of that anyway...
